### ISSUE TITLE
Gradient accumulation causing different training curves

### ISSUE THREAD

**Issue by ZhaofengWu** [huggingface/transformers#14638]
For a pretraining experiment using the built-in `Trainer`, setting batch size 32 x 16 accumulation steps seems to yield a different training loss curve from 64 x 8. Even though they converge to approximately the same value, shouldn't the curves be exactly the same? What would cause the difference?
![image](https://user-images.githubusercontent.com/11954789/144804002-fbd31936-c58a-45c4-a991-5eabcde42f16.png)

I've also seen similar things with using 1 vs. multiple GPUs (under DataParallel), which may or may not be relevant.

**Comment by gugarosa** [comment]
Theoretically speaking, it is expected because each configuration differently feed the gradients. Though is also expected that they converge to a common point.

**Comment by ZhaofengWu** [comment]
What do you mean by they differently feed the gradients? Calculating the loss based on 32 examples and then another 32 examples and then calculating the gradients should be exactly the same as calculating the loss based on 64 examples and then calculating the gradients, no?

**Comment by gugarosa** [comment]
They are very almost the same. Though, they are being repeatedly calculated and might lead to some numerical difference that explain the divergence at some points

**Comment by LysandreJik** [comment]
cc @sgugger

**Comment by sgugger** [comment]
I think @gugarosa analysis is correct in the sense that it could be explained by precision differences when computing the gradients vs accumulating partial gradients.

**Comment by ZhaofengWu** [comment]
Do you think that is suffcient to explain the not-too-minor difference in the plot?

**Comment by sgugger** [comment]
It's hard to say given you shared basically nothing on the training script :-)

**Comment by ZhaofengWu** [comment]
Ok, I think I'll take this explanation. Thanks! I'm not doing anything extraordinary in my training script and posting it in its entirety will probably bore everyone :)

**Comment by ZhaofengWu** [comment]
@sgugger I found one source of difference: the grouping by length done in `LengthGroupedSampler` depends on the batch size, so the data ordering would be different. I argue that ideally in all `(batch size, gradient accumulation)` settings, as long as the effective batch size is the same, they should have an identical training process, except for numerical issues.

**Comment by sgugger** [comment]
Good point, will try to see if there is a quick way to fix this!

**Comment by HuangChiEn** [comment]
> [@sgugger](https://github.com/sgugger) I found one source of difference: the grouping by length done in `LengthGroupedSampler` depends on the batch size, so the data ordering would be different. I argue that ideally in all `(batch size, gradient accumulation)` settings, as long as the effective batch size is the same, they should have an identical training process, except for numerical issues.

hello, i wonder why the different order for feeding data will change the accumulated gradient ?

**Comment by siridris** [comment]
look like unsloth have a fix for this 
https://unsloth.ai/blog/gradient

**Comment by HuangChiEn** [comment]
> look like unsloth have a fix for this https://unsloth.ai/blog/gradient

wow, that's a good news ~
